### PR TITLE
Nodes config init

### PR DESCRIPTION
**Pull request by AhmedSoliman**
Nodes config init

Introduces a basic serializable structure for a versioned nodes configuration.
This is currently unused and have structures that will be merged with changes from #1151 when merged

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1168).
* #1172
* #1171
* __->__ #1168
* #1166
* #1164

### CODE REVIEW

File: crates/types/src/nodes_config.rs [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+enum MaybeNode {
+    Tombstone,
+    Node(NodeConfig),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+struct NodeConfig {
+    pub name: String,
```
Comment (tillrohrmann) [inline crates/types/src/nodes_config.rs:76]: Are names a convenience feature for users to specify a name for a node?

**Review by tillrohrmann — APPROVED** [review]
Nice work :-) LGTM. +1 for merging. I will try to make use of the `NodeConfig` when starting a Restate server.

File: crates/types/src/nodes_config.rs (hunk 1 above)
Comment (AhmedSoliman) [reply crates/types/src/nodes_config.rs:76]: node names are what the user actually provides on startup and attach should return a generational node id. Environments like k8s will give us a stable name which controller can use to locate the node and bump the generation on attach. I'll make the change in a follow up commit to do that.

Comment (tillrohrmann) [reply crates/types/src/nodes_config.rs:76]: Alright. So far I was assuming that the `NodeId` would be the stable identifier for a node.

**Pull request merged**